File: 0003_Longest_Substring_Without_Repeating_Characters/Longest_Substring_Without_Repeating_Characters.cpp
// Check py file for the explaining comments

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int r = 0;
        int l = 0;
        int maxlen = 0;
        unordered_map<char, int> dc = {};
        for(r; r < s.size(); r++){
            char ch = s[r];
            if (dc.count(ch)){
                if (l <= dc[ch]){         
                    l = dc[ch] + 1;
                }   
            }

            dc[ch] = r;
            int rllength = r - l + 1;
            maxlen = max(rllength, maxlen);
        }

        return maxlen;
    }
};

File: 0003_Longest_Substring_Without_Repeating_Characters/Longest_Substring_Without_Repeating_Characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        r = 0
        l = 0
        maxlen = 0
        dc = {}
        for r, ch in enumerate(s):
            if ch in dc.keys():
                if l <= dc[ch]:         
                    l = dc[ch] + 1
                    dc[ch] = r
                else:
                    dc[ch] = r

            else:
                dc[ch] = r
            rllength = r - l + 1
            maxlen = max(rllength, maxlen)

        return maxlen